### PR TITLE
8341471: Reversed field layout caused by unstable sorting

### PR DESCRIPTION
**Pull request by fg1417**
For class Test:
```
    public class Test {
        char a000;
        char a001;
        char a002;
        char a003;
        char a004;
        char a005;
        char a006;
        char a007;
        char a008;
        char a009;
        char a00a;
        char a00b;
    }
```

We found its field layout on macOS was:

```
Layout of class Test
  Instance fields:
   @0 12/- RESERVED
   @12 "a00b" C 2/2 REGULAR
   @14 "a001" C 2/2 REGULAR
   @16 "a002" C 2/2 REGULAR
   @18 "a003" C 2/2 REGULAR
   @20 "a004" C 2/2 REGULAR
   @22 "a005" C 2/2 REGULAR
   @24 "a006" C 2/2 REGULAR
   @26 "a007" C 2/2 REGULAR
   @28 "a008" C 2/2 REGULAR
   @30 "a009" C 2/2 REGULAR
   @32 "a00a" C 2/2 REGULAR
   @34 "a000" C 2/2 REGULAR
```

`a000` was put in the end while `a00b` was put in the beginning.

Fields get sorted according to size in [1]. `qsort()` on macOS reverses the order of fields with the same size. We should extend the comparison function to preserve the order on macOS, as we did on Windows.

Tier 1-3 passed on macOS.

[1] https://github.com/openjdk/jdk/blob/3ee94e040a7395d11a294a6b660d707c97f188f8/src/hotspot/share/classfile/fieldLayoutBuilder.cpp#L102

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341471](https://bugs.openjdk.org/browse/JDK-8341471): Reversed field layout caused by unstable sorting (**Bug** - P4)

### Reviewers
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21382/head:pull/21382` \
`$ git checkout pull/21382`

Update a local copy of the PR: \
`$ git checkout pull/21382` \
`$ git pull https://git.openjdk.org/jdk.git pull/21382/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21382`

View PR using the GUI difftool: \
`$ git pr show -t 21382`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21382.diff">https://git.openjdk.org/jdk/pull/21382.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21382#issuecomment-2396332991)

### CODE REVIEW

**Review by TheShermanTanker — APPROVED** [review]
Not usually my domain, but this looks good since it's not that big of a change

File: src/hotspot/share/classfile/fieldLayoutBuilder.hpp [hunk 1]
```diff
@@ -101,8 +101,8 @@ class LayoutRawBlock : public ResourceObj {
   // sort fields in decreasing order.
   // Note: with line types, the comparison should include alignment constraint if sizes are equals
   static int compare_size_inverted(LayoutRawBlock** x, LayoutRawBlock** y)  {
-#ifdef _WINDOWS
-    // qsort() on Windows reverse the order of fields with the same size
+#if defined (_WINDOWS) || defined (__APPLE__)
+    // qsort() on Windows/macOS reverse the order of fields with the same size
     // the extension of the comparison function below preserves this order
```
Comment (theRealAph) [inline src/hotspot/share/classfile/fieldLayoutBuilder.hpp:106]: `qsort()` isn't defined to be stable anywhere. I can't see any reason to restrict thus bug fix to Windows and MacOS.

Comment (fg1417) [reply src/hotspot/share/classfile/fieldLayoutBuilder.hpp]: @theRealAph thanks for your review. The new commit removed the macro and applied the extension of the comparison function to all platforms. Tier 1 - 3 passed on both linux-aarch64 and x86 platforms.

**Review by jdksjolen — COMMENTED** [review]
Please add a comment that mentions the reason for the branch, for example

```c++
// ensure stable sort
if (...) {

}
```

**Comment by fg1417** [comment]
> Please add a comment that mentions the reason for the branch, for example
> 
> ```c++
> // ensure stable sort
> if (...) {
> 
> }
> ```

Thanks for your review @jdksjolen . Updated it with comments.

**Review by jdksjolen — APPROVED** [review]
LGTM

**Comment by fg1417** [comment]
May I have a review for the latest commit? @theRealAph @TheShermanTanker

Thanks!

**Review by TheShermanTanker — APPROVED** [review]
In my opinion it was sufficient to use it only for macOS and Windows, but I don't mind it for all platforms

**Comment by fg1417** [comment]
Thanks for your review @theRealAph @TheShermanTanker @jdksjolen .

I'll integrate it.

/integrate

**Pull request closed without merging**